### PR TITLE
Search fields in custom POI filter

### PR DESCRIPTION
**Pull request by Skalii**

### CODE REVIEW

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 1]
```diff
@@ -145,6 +268,79 @@ - (IBAction)onShowButtonClicked:(id)sender
...
+    if (textView.text.length == 0)
+    {
+        _searchMode = NO;
+        self.saveButton.hidden = NO;
+        [self.tableView setEditing:NO];
+        self.tableView.allowsMultipleSelectionDuringEditing = NO;
+        self.tableView.tableHeaderView = [OAUtilities setupTableHeaderViewWithText:OALocalizedString(@"search_poi_types_descr") font:kHeaderViewFont textColor:UIColorFromRGB(color_text_footer) lineSpacing:6.0 isTitle:NO];
+        [_core updateSettings:_core.getSearchSettings.resetSearchTypes];
+    }
+    else {
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:305]: Please be more consistent with braces formatting.

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 2]
```diff
@@ -145,6 +268,79 @@ - (IBAction)onShowButtonClicked:(id)sender
...
+                    if ([poiObject isKindOfClass:[OAPOIType class]])
+                    {
+                        OAPOIType *poiType = (OAPOIType *) poiObject;
+                        if (!poiType.isAdditional)
+                            [results addObject:poiType];
+                    }
+                }
+                dispatch_async(dispatch_get_main_queue(), ^{
+                    _searchResult = [NSMutableArray arrayWithArray:results];
+                    [self.tableView reloadSections:[NSIndexSet indexSetWithIndex:0] withRowAnimation:UITableViewRowAnimationNone];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:331]: Consider using automatic animation when possible.

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 3]
```diff
@@ -171,59 +368,139 @@ - (void)selectSubcategoryDone:(OAPOICategory *)category keys:(NSMutableSet<NSStr
...
+
+        if (cell) {
+            OAPOICategory *category = _categories[indexPath.row];
+            NSSet<NSString *> *subtypes = [_acceptedTypes objectForKey:category];
+            NSInteger countAcceptedTypes = subtypes.count;
+            NSInteger countAllTypes = category.poiTypes.count;
+            BOOL isSelected = [_filter isTypeAccepted:category] && subtypes.count > 0;
+            cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
+
+            cell.contentView.backgroundColor = [UIColor whiteColor];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:386]: Seems redundant here.

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 4]
```diff
@@ -171,59 +368,139 @@ - (void)selectSubcategoryDone:(OAPOICategory *)category keys:(NSMutableSet<NSStr
...
+            cell.separatorInset = UIEdgeInsetsMake(0.0, 66.0, 0.0, 0.0);
+
+            cell.textView.text = category.nameLocalized;
+            cell.textView.textColor = [UIColor blackColor];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:390]: Seems redundant

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 5]
```diff
@@ -171,59 +368,139 @@ - (void)selectSubcategoryDone:(OAPOICategory *)category keys:(NSMutableSet<NSStr
...
+            cell.imgView.image = img;
+            cell.imgView.tintColor = isSelected ? UIColorFromRGB(color_primary_purple) : UIColorFromRGB(color_tint_gray);
+            cell.imgView.contentMode = UIViewContentModeCenter;
+
+            NSString *descText;
+            if (subtypes == [OAPOIBaseType nullSet] || countAllTypes == countAcceptedTypes)
+                descText = [NSString stringWithFormat:@"%@ - %lu", OALocalizedString(@"shared_string_all"), countAllTypes];
+            else
+                descText = [NSString stringWithFormat:@"%lu/%lu", countAcceptedTypes, countAllTypes];
+            cell.descriptionView.hidden = false;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:402]: In obj c, we use YES/NO for boolean values (do not confuse with c/c++ true/false)

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 6]
```diff
@@ -171,59 +368,139 @@ - (void)selectSubcategoryDone:(OAPOICategory *)category keys:(NSMutableSet<NSStr
...
+            cell.descriptionView.text = descText;
+            cell.descriptionView.textColor = UIColorFromRGB(color_text_footer);
+
+            [cell updateConstraints];
+            return cell;
+        }
+    } else {
+        OAMenuSimpleCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellTypeTitle];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:410]: Confusing code, why do you create the same cell twice? You should create it once and only then adjust the data according to the _searchMode flag.

File: Sources/Controllers/2.0/OASelectSubcategoryViewController.mm [hunk 7]
```diff
@@ -92,33 +100,68 @@ - (void)viewDidLoad
...
 -(void)applyLocalization
 {
+    [self updateTextTitle];
+
     self.applyButton.titleLabel.text = OALocalizedString(@"shared_string_apply");
+    self.cancelSearchButton.titleLabel.text = OALocalizedString(@"shared_string_cancel");
 }
 
-- (IBAction)onBackButtonClicked:(id)sender
+- (void)updateTextTitle
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OASelectSubcategoryViewController.mm:118]: Better to rename it to updateScreenTitle

File: Sources/Controllers/2.0/OASelectSubcategoryViewController.mm [hunk 8]
```diff
@@ -92,33 +100,68 @@ - (void)viewDidLoad
...
-
-        [_delegate selectSubcategoryDone:_category keys:selectedKeys allSelected:_selectedItems.count == _items.count];
+    if (_searchMode) {
+        self.bottomView.hidden = YES;
+        self.applyButton.hidden = YES;
+        self.tableBottomConstraint.constant = 0;
+    } else {
+        self.bottomView.hidden = NO;
+        self.applyButton.hidden = NO;
+        self.tableBottomConstraint.constant = 87;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OASelectSubcategoryViewController.mm:137]: Again, please mind the safe area.

File: Sources/Controllers/2.0/OASelectSubcategoryViewController.mm [hunk 9]
```diff
@@ -92,33 +100,68 @@ - (void)viewDidLoad
...
     }
+}
 
-    [self.navigationController popViewControllerAnimated:YES];
+- (void)updateSearchView:(BOOL)searchMode
+{
+    UIView *searchLeftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 30, 20)];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OASelectSubcategoryViewController.mm:143]: Copy/paste code

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 10]
```diff
@@ -368,29 +376,35 @@ - (void)selectSubcategoryDone:(OAPOICategory *)category keys:(NSMutableSet<NSStr
...
-        OAMenuSimpleCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellTypeTitleDescCollapse];
-        if (cell == nil) {
-            NSArray *nib = [[NSBundle mainBundle] loadNibNamed:kCellTypeTitleDescCollapse owner:self options:nil];
-            cell = (OAMenuSimpleCell *) nib[0];
+    OAMenuSimpleCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellTypeTitleDescCollapse];
+    if (cell == nil)
+    {
+        NSArray *nib = [[NSBundle mainBundle] loadNibNamed:kCellTypeTitleDescCollapse owner:self options:nil];
+        cell = (OAMenuSimpleCell *) nib[0];
+        cell.separatorInset = UIEdgeInsetsMake(0.0, 65.0, 0.0, 0.0);
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:392]: Will this display correctly after screen rotation?

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 11]
```diff
@@ -186,6 +186,11 @@ - (void)updateTextShowButton
 }
 
 - (void)updateSearchView:(BOOL)searchMode
+{
+    [OACustomPOIViewController updateSearchView:searchMode search:self.searchField cancel:self.cancelSearchButton rightConstraint:self.searchFieldRightConstraint];
+}
+
++ (void)updateSearchView:(BOOL)searchMode search:(UITextField *)search cancel:(UIButton *)cancel rightConstraint:(NSLayoutConstraint *)rightConstraint;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:193]: Prefer naming variables for what they are e.g. searchField, cancelButton

File: Sources/Controllers/2.0/OACustomPOIViewController.h [hunk 12]
```diff
@@ -19,4 +19,6 @@
 
 - (instancetype)initWithFilter:(OAPOIUIFilter *)filter;
 
++ (void)updateSearchView:(BOOL)searchMode searchField:(UITextField * _Nonnull)searchField cancelButton:(UIButton * _Nonnull)cancelButton searchFieldRightConstraint:(NSLayoutConstraint * _Nonnull)searchFieldRightConstraint;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.h:22]: What is the purpose of this method? I don't even see its implementation anywhere.

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 13]
```diff
@@ -15,29 +15,42 @@
...
+#import "OAMenuSimpleCell.h"
+#import "OASearchUICore.h"
+#import "OAQuickSearchHelper.h"
+#import "OASearchSettings.h"
+#import "OAPOIType.h"
+#import "OAPOIFilterViewController.h"
+#import "OATableViewCustomHeaderView.h"
 
 #define kHeaderViewFont [UIFont systemFontOfSize:15.0]
+#define kHeaderId @"TableViewSectionHeader"
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:28]: Migate to static id method

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 14]
```diff
@@ -15,29 +15,42 @@
...
 
 @property (weak, nonatomic) IBOutlet UIView *navBar;
 @property (weak, nonatomic) IBOutlet UITableView *tableView;
 @property (weak, nonatomic) IBOutlet UILabel *titleLabel;
 @property (weak, nonatomic) IBOutlet UIButton *backButton;
 @property (weak, nonatomic) IBOutlet UIButton *saveButton;
 @property (weak, nonatomic) IBOutlet UIButton *showButton;
+@property (weak, nonatomic) IBOutlet UISearchBar *searchBar;
+@property (weak, nonatomic) IBOutlet UIView *bottomView;
+@property (weak, nonatomic) IBOutlet NSLayoutConstraint *tableBottomConstraint;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:40]: All outlets to constraints should be strong.

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 15]
```diff
@@ -68,39 +99,77 @@ - (void)viewDidLoad
 
     self.tableView.delegate = self;
     self.tableView.dataSource = self;
-    self.tableView.tableHeaderView = [OAUtilities setupTableHeaderViewWithText:OALocalizedString(@"search_poi_types_descr") font:kHeaderViewFont textColor:UIColorFromRGB(color_text_footer) lineSpacing:6.0 isTitle:NO];
+    [self.tableView registerClass:OATableViewCustomHeaderView.class forHeaderFooterViewReuseIdentifier:kHeaderId];
+
+    _searchMode = NO;
+    self.searchBar.delegate = self;
+    UITextField *textField = [self.searchBar valueForKey:@"searchField"];
+    textField.backgroundColor = [UIColor colorWithWhite:1 alpha:1];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:107]: ```UIColor.whiteColor```

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 16]
```diff
@@ -68,39 +99,77 @@ - (void)viewDidLoad
 
     self.tableView.delegate = self;
     self.tableView.dataSource = self;
-    self.tableView.tableHeaderView = [OAUtilities setupTableHeaderViewWithText:OALocalizedString(@"search_poi_types_descr") font:kHeaderViewFont textColor:UIColorFromRGB(color_text_footer) lineSpacing:6.0 isTitle:NO];
+    [self.tableView registerClass:OATableViewCustomHeaderView.class forHeaderFooterViewReuseIdentifier:kHeaderId];
+
+    _searchMode = NO;
+    self.searchBar.delegate = self;
+    UITextField *textField = [self.searchBar valueForKey:@"searchField"];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:106]: Make sure this is not prohibited by Apple and will not get us banned from the appstore

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 17]
```diff
@@ -68,39 +99,77 @@ - (void)viewDidLoad
...
+- (void)updateShowButton:(BOOL)hasSelection
+{
+    self.showButton.backgroundColor = hasSelection ? UIColorFromRGB(color_primary_purple) : UIColorFromRGB(color_route_button_inactive);
+    [self.showButton setTintColor:hasSelection ? UIColor.whiteColor : UIColorFromRGB(color_text_footer)];
+    [self.showButton setTitleColor:hasSelection ? UIColor.whiteColor : UIColorFromRGB(color_text_footer) forState:UIControlStateNormal];
+    [self.showButton setUserInteractionEnabled:hasSelection];
+
+    if (_searchMode)
+    {
+        self.bottomView.hidden = YES;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:139]: Simplify by directly setting the bool to hidden property and using a ternary expression for ```tableBottomConstraint.constant```

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 18]
```diff
@@ -68,39 +99,77 @@ - (void)viewDidLoad
...
-    NSMutableAttributedString *attrShow = [[NSMutableAttributedString alloc] initWithString:textShow attributes:@{NSFontAttributeName:fontShow, NSForegroundColorAttributeName:colorShow}];
+    UIColor *colorShow = _countShowCategories != 0 ? UIColor.whiteColor : UIColorFromRGB(color_text_footer);
+    NSMutableAttributedString *attrShow = [[NSMutableAttributedString alloc] initWithString:textShow attributes:@{NSFontAttributeName: fontShow, NSForegroundColorAttributeName: colorShow}];
 
     NSString *textCategories = [NSString stringWithFormat:@"\n%@: %li", OALocalizedString(@"categories"), _countShowCategories];
     UIFont *fontCategories = [UIFont systemFontOfSize:13 weight:UIFontWeightRegular];
-    UIColor *colorCategories = [[UIColor alloc] initWithWhite:1 alpha:0.5];
-    NSMutableAttributedString *attrCategories = [[NSMutableAttributedString alloc] initWithString:textCategories attributes:@{NSFontAttributeName:fontCategories, NSForegroundColorAttributeName:colorCategories}];
+    UIColor *colorCategories = _countShowCategories != 0 ? [[UIColor alloc] initWithWhite:1 alpha:0.5] : UIColorFromRGB(color_text_footer);
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:171]: Is [UIColor alloc] initWithWhite:1 alpha:0.5] from Figma?

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 19]
```diff
@@ -109,7 +178,45 @@ - (void)updateTextShowButton
...
     [attrShow addAttribute:NSParagraphStyleAttributeName value:style range:NSMakeRange(0, attrShow.string.length)];
 
-    [_showButton setAttributedTitle:attrShow forState:UIControlStateNormal];
+    [self.showButton setAttributedTitle:attrShow forState:UIControlStateNormal];
+    [self updateShowButton:_countShowCategories != 0];
+}
+
+- (NSString *)getTitleForSection
+{
+    return OALocalizedString(@"search_poi_types_descr");
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:187]: Not sure this method is needed at all

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 20]
```diff
@@ -161,79 +336,163 @@ - (void)selectSubcategoryDone:(OAPOICategory *)category keys:(NSMutableSet<NSStr
...
 {
-    return [OAPOISearchHelper getHeightForFooter];
+    return _searchMode;
 }
 
-- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
+- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
 {
-    return [OAPOISearchHelper getHeightForHeader];
+    return _searchMode ? 48 : 66;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:471]: Again, do not implement ```heightForRowAtIndexPath``` under any circumstance except for non-dynamic cells (e.g. picker cells that have a constant height). In the autolayout context all the cell heights are calculated dynamically by the system.

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 21]
```diff
@@ -161,79 +336,163 @@ - (void)selectSubcategoryDone:(OAPOICategory *)category keys:(NSMutableSet<NSStr
...
+}
+
+- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
+{
+    if (section == 0 && !_searchMode) {
+        OATableViewCustomHeaderView *customHeader = [tableView dequeueReusableHeaderFooterViewWithIdentifier:kHeaderId];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:477]: Remove kHeaderId

File: Sources/Controllers/2.0/OACustomPOIViewController.mm [hunk 22]
```diff
@@ -161,79 +336,163 @@ - (void)selectSubcategoryDone:(OAPOICategory *)category keys:(NSMutableSet<NSStr
...
+}
+
+- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
+    if (section == 0 && !_searchMode) {
+        return [OATableViewCustomHeaderView getHeight:[self getTitleForSection] width:tableView.bounds.size.width] + 24;
+    }
+    return 0.01;
+}
+
+- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OACustomPOIViewController.mm:493]: Set this in the xib and do not call this method. Also, please check other places OAPOISearchHelper is used, and remove this useless class from the project entirely.

File: Sources/Controllers/2.0/OADeleteCustomFiltersViewController.mm [hunk 23]
```diff
@@ -125,14 +125,13 @@ - (void)setupDeleteButtonView
 
 - (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath
 {
-    NSObject *item = _items[indexPath.row];
     NSString *cellType = indexPath.row == 0 ? [OACustomSelectionButtonCell getCellIdentifier] : [OAMenuSimpleCell getCellIdentifier];
     if ([cellType isEqualToString:[OACustomSelectionButtonCell getCellIdentifier]])
     {
-        OACustomSelectionButtonCell *cell = [tableView dequeueReusableCellWithIdentifier:[OACustomSelectionButtonCell getCellIdentifier]];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OADeleteCustomFiltersViewController.mm:132]: Please do not regress the code to the incorrect state. Always prefer code from master when fixing the conflicts.

File: Sources/Controllers/2.0/OASelectSubcategoryViewController.mm [hunk 24]
```diff
@@ -15,30 +15,44 @@
...
 #import "OAColors.h"
 #import "OAPOIUIFilter.h"
+#import "OASearchResult.h"
+#import "OASearchUICore.h"
+#import "OASearchSettings.h"
+#import "OAQuickSearchHelper.h"
+#import "OATableViewCustomHeaderView.h"
 
-@interface OASelectSubcategoryViewController () <UITableViewDataSource, UITableViewDelegate, OAMultiselectableHeaderDelegate>
+#define kHeaderId @"TableViewSectionHeader"
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OASelectSubcategoryViewController.mm:24]: Remove this definition and use the correct way of getting the identifier

File: Sources/Controllers/2.0/OASelectSubcategoryViewController.mm [hunk 25]
```diff
@@ -55,67 +69,101 @@ - (void)initData
...
-        NSMutableSet<NSString *> *selectedKeys = [NSMutableSet set];
-        for (OAPOIType *poiType in _selectedItems)
-            [selectedKeys addObject:poiType.name];
+    self.applyButton.backgroundColor = hasSelection ? UIColorFromRGB(color_primary_purple) : UIColorFromRGB(color_route_button_inactive);
+    [self.applyButton setTintColor:hasSelection ? UIColor.whiteColor : UIColorFromRGB(color_text_footer)];
+    [self.applyButton setTitleColor:hasSelection ? UIColor.whiteColor : UIColorFromRGB(color_text_footer) forState:UIControlStateNormal];
+    [self.applyButton setUserInteractionEnabled:hasSelection];
 
-        [_delegate selectSubcategoryDone:_category keys:selectedKeys allSelected:_selectedItems.count == _items.count];
+    if (_searchMode)
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OASelectSubcategoryViewController.mm:142]: Simplify as mentioned in the previous class

File: Sources/Controllers/2.0/OACustomPOIViewController.mm (hunk 18 above)
Comment (Skalii) [reply Sources/Controllers/2.0/OACustomPOIViewController.mm:171]: Yes

**Pull request merged**